🔋 Enhance cupertino button fade in and fade out

YazeedAlKhalaf opened the pull request:
This part of the combing process of Cupertino widgets to make sure they match the swift ui implementation.

| Before | After |
| ------ | ------ |
| <video src="https://user-images.githubusercontent.com/41989385/127230200-27633427-64db-48da-b0b2-e1e400175671.mov" /> | <video src="https://user-images.githubusercontent.com/41989385/127230203-3ae3286c-2d8a-456e-b513-057867bcecfe.mov" /> |

helps in fixing #72545 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

YazeedAlKhalaf committed:
🔋 Enhance cupertino button fade in and fade out

jmagman commented:
There's a failing test:
```
01:23 +2358 ~13: /opt/s/w/ir/x/w/flutter/packages/flutter/test/cupertino/dialog_test.dart: FadeTransition animation for showCupertinoDialog()                                                          
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure object was thrown running a test:
  Expected: 0.4 (±0.001)
  Actual: <0.46496927734375004>
   Which: 0.46496927734375004 is not in the range of 0.4 (±0.001).

When the exception was thrown, this was the stack:
#4      main.<anonymous closure> (file:///opt/s/w/ir/x/w/flutter/packages/flutter/test/cupertino/dialog_test.dart:1046:5)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

This was caught by the test expectation on the following line:
  file:///opt/s/w/ir/x/w/flutter/packages/flutter/test/cupertino/dialog_test.dart line 1046
The test description was:
  FadeTransition animation for showCupertinoDialog()
════════════════════════════════════════════════════════════════════════════════════════════════════

01:23 +2358 ~13 -1: /opt/s/w/ir/x/w/flutter/packages/flutter/test/cupertino/dialog_test.dart: FadeTransition animation for showCupertinoDialog() [E]                                                   
  Test failed. See exception logs above.
  The test description was: FadeTransition animation for showCupertinoDialog()
```

jmagman commented:
What iOS version are you testing?  iOS 14?

YazeedAlKhalaf commented:
> What iOS version are you testing?  iOS 14?

iOS 15 in using XCode 13 and iOS 15 on a real device.

YazeedAlKhalaf commented:
> There's a failing test:
> 
> ```
> 
> 01:23 +2358 ~13: /opt/s/w/ir/x/w/flutter/packages/flutter/test/cupertino/dialog_test.dart: FadeTransition animation for showCupertinoDialog()                                                          
> 
> ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
> 
> The following TestFailure object was thrown running a test:
> 
>   Expected: 0.4 (±0.001)
> 
>   Actual: <0.46496927734375004>
> 
>    Which: 0.46496927734375004 is not in the range of 0.4 (±0.001).
> 
> 
> 
> When the exception was thrown, this was the stack:
> 
> #4      main.<anonymous closure> (file:///opt/s/w/ir/x/w/flutter/packages/flutter/test/cupertino/dialog_test.dart:1046:5)
> 
> <asynchronous suspension>
> 
> <asynchronous suspension>
> 
> (elided one frame from package:stack_trace)
> 
> 
> 
> This was caught by the test expectation on the following line:
> 
>   file:///opt/s/w/ir/x/w/flutter/packages/flutter/test/cupertino/dialog_test.dart line 1046
> 
> The test description was:
> 
>   FadeTransition animation for showCupertinoDialog()
> 
> ════════════════════════════════════════════════════════════════════════════════════════════════════
> 
> 
> 
> 01:23 +2358 ~13 -1: /opt/s/w/ir/x/w/flutter/packages/flutter/test/cupertino/dialog_test.dart: FadeTransition animation for showCupertinoDialog() [E]                                                   
> 
>   Test failed. See exception logs above.
> 
>   The test description was: FadeTransition animation for showCupertinoDialog()
> 
> ```

Failed locally, I re ran it locally and it succeeded. Weird 🙈.

YazeedAlKhalaf commented:
Using iOS 15 because it will be released soon to the public as the stable version.

jmagman commented:
This can't be merged until the tests pass.

jmagman reviewed (changes requested):
I believe there's a bug in the failing test.  It was incorrectly checking the `FadeTransition` of the button instead of the Cuptertino dialog.  Once https://github.com/flutter/flutter/pull/87533 please rebase, and that test should pass.

This PR also needs a test.  There wasn't a unit test for button animation already, but I added one in https://github.com/flutter/flutter/pull/87535.  Contributing PRs is as much about writing and updating tests as it is about adding or updating a feature.  🙂 

Once that PR merges, you can update that test for the updated animation.

YazeedAlKhalaf commented:
> I believe there's a bug in the failing test.  It was incorrectly checking the `FadeTransition` of the button instead of the Cuptertino dialog.  Once https://github.com/flutter/flutter/pull/87533 please rebase, and that test should pass.
> 
> 
> 
> This PR also needs a test.  There wasn't a unit test for button animation already, but I added one in https://github.com/flutter/flutter/pull/87535.  Contributing PRs is as much about writing and updating tests as it is about adding or updating a feature.  🙂 
> 
> 
> 
> Once that PR merges, you can update that test for the updated animation.

Sure thanks 🚀

jmagman commented:
Both tests have merged. Please rebase and update the new button animation test.

YazeedAlKhalaf commented:
> Both tests have merged. Please rebase and update the new button animation test.

will do now 🚀

YazeedAlKhalaf committed:
✏️ Edit button_test of cupertino to make new animation duration pass the test

YazeedAlKhalaf commented:
flutter-gold is pending but others are successful.
I am not sure what flutter-gold is, can tell me? 😁

jmagman commented on packages/flutter/test/cupertino/button_test.dart:
```diff
@@ -205,21 +205,53 @@ void main() {
     await tester.pump();
     FadeTransition transition = tester.firstWidget(find.byType(FadeTransition));
 
+    await tester.pump(const Duration(milliseconds: 25));
```
Can you make these 50ms increments? 25 is a bit granular for this.

YazeedAlKhalaf replied:
sure 🚀

jmagman reviewed (commented):
Minor nit about reducing the size of the test.  When that's done it'll be ready to merge!

> EDIT: everything is successful 🚀, but still don't know what flutter-gold is 🙋‍♂️.

https://medium.com/flutter-community/flutter-golden-tests-compare-widgets-with-snapshots-27f83f266cea
https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter

Golden tests have passed.

YazeedAlKhalaf committed:
🔋 Enhance button_test

- started using 50 milliseconds intervals instead of 25 milliseconds.
- makes te test code less.
- one of the pump intervals is 100 because the value does not change due to the curve used in the button animation.

jmagman reviewed (approved):
LGTM, thanks for the contribution!
